### ISSUE TITLE
Unable to crane copy from docker.elastic.co

### ISSUE THREAD

**Issue by Roshick** [google/go-containerregistry#1108]
Hello,

I'm not sure if this is actually a crane problem, but when trying to `crane copy` or  `crane pull` any image from "docker.elastic.co" I get a "https://docker.elastic.co/v2/: read tcp ip -> ip: read: connection reset by peer" (example image: "docker.elastic.co/apm/apm-server:7.7.1").

What I noticed is that https://docker.elastic.co/v2/ appears to require authorization and results in a 401, while https://docker.elastic.co returns a 200.

`docker pull` is actually able to pull the image, I guess it probably tries to use v2 first and then goes for v1, where it is able to pull?!

Thanks for any help or idea how I can fix this issue! :)

**Comment by jonjohnsonjr** [comment]
Works for me :/

I'm guessing you have some proxy/firewall preventing you from talking to elastic, and docker [has been configured](https://docs.docker.com/network/proxy/) to use this proxy (or the daemon is running on a separate machine that can access elastic?).

> appears to require authorization and results in a 401

This is normal, e.g. docker hub also requires you to authenticate for public images.

**Comment by imjasonh** [comment]
fwiw this didn't work for me using `crane cp`:

```
$ crane cp docker.elastic.co/apm/apm-server:7.7.1 gcr.io/imjasonh/elastic --verbose
...things proceed quickly, then stop

2021/08/16 13:28:26 <-- 200 https://d2iks1dkcwqcbx.cloudfront.net/docker/registry/v2/blobs/sha256/9c/9cb05eb8b19003f2e96f1dcb75ffe5c1d01fd278f9f064ce8fbf00028bd85585/data?Expires=1629136100&Signature=Pq0hZKRG9Gjpwm2qvfPALMi~JGuNTktxOPAqGMqybQZEj9wBcCnDPfFVvkNaDkakLqH8-PeoL1EiUITPQ5TJeT5mxEuFizprLJIms7y7ExcQl4L5khJ7bws-tVk4T9o7E2M8jXU-eB8UV8iJ11yQsC9U56vjLK7tiO8F9NWp3ygu31p9lwtBiXafiw7ouFVY-t4ULkpoZT0N6KZBn0HzsECJB1WMMZA5yiaOPAl-Xg3GEMXTfTYXC1d60fDRlKZr325Tdoww9cOAOVdaIIdWCvmy4dBUDDSLPHoXEocOgz0UedUecZ0Qx3lt0WcmJ3IEd-~Ou4Xnh0ZA8owQsm3AhQ__&Key-Pair-Id=APKAIVAVKHB6SNHJAJQQ (5.830010262s) [body redacted: omitting binary blobs from logs]
2021/08/16 13:28:26 HTTP/1.1 200 OK
Content-Length: 26061111
Accept-Ranges: bytes
Connection: keep-alive
Content-Type: application/octet-stream
Date: Mon, 16 Aug 2021 17:28:22 GMT
Etag: "7b508e9f38541af72b71a331d04c2f5d"
Last-Modified: Thu, 28 May 2020 19:40:48 GMT
Server: AmazonS3
Via: 1.1 0a84f3fd72c51aed32833055c4b982d3.cloudfront.net (CloudFront)
X-Amz-Cf-Id: 7RO505NZBcUdhDCUT6ml-TLh88NX6jq4OBvln_WLY4xnvgDrhqryQw==
X-Amz-Cf-Pop: BOS50-C3
X-Amz-Server-Side-Encryption: AES256
X-Cache: Miss from cloudfront


2021/08/16 13:28:26 --> PATCH https://gcr.io/v2/imjasonh/elastic/blobs/uploads/AJAD5v13to2CaoVvJcqVMaSZAgX0_m-7JxbvTqOwVnpqXoYjxW1AuRFJrd0BF951w9cs6Jcq9S94F5zexVrgeAs [body redacted: omitting binary blobs from logs]
2021/08/16 13:28:26 PATCH /v2/imjasonh/elastic/blobs/uploads/AJAD5v13to2CaoVvJcqVMaSZAgX0_m-7JxbvTqOwVnpqXoYjxW1AuRFJrd0BF951w9cs6Jcq9S94F5zexVrgeAs HTTP/1.1
Host: gcr.io
User-Agent: crane/0.6.0 go-containerregistry/(devel)
Transfer-Encoding: chunked
Authorization: <redacted>
Content-Type: application/octet-stream
Accept-Encoding: gzip

# ...hangs, ctrl+c

^C2021/08/16 13:31:35 <-- context canceled PATCH https://gcr.io/v2/imjasonh/elastic/blobs/uploads/AJAD5v2lhllJ7p8DZ3y0TwZ6IV66iLJEPoDSkXPUzimmdaOBK_GprROAINc-RIGDgT-D14iEf4sSgw2eU1EKz5A (3m14.847169312s)
2021/08/16 13:31:35 <-- context canceled GET https://d2iks1dkcwqcbx.cloudfront.net/docker/registry/v2/blobs/sha256/4e/4e7de3b5d9692ba589736384a43de0c601a1573b10b25d4a10eb513170e7553f/data?Expires=1629136100&Signature=auFg9WR5buvRczfvlNiva46HUU7-P1ZHD1R50sihjEe2vXZEomAC1whXt7NemcIHR6V2GBPHXC2gxiJTMMB~swA7CsGNGPTLc2mF-yBtAkJ2Dm6sqxiAZo5UkBmMFSKrZbHIG2BJefoaQZdRXXQHqXx-4nWGKJ670A0qjvNgfSgdThwgwNn3qVIUmcGMrvZnPebf4J7xUdZlXVjuiJLE14sP8TlLOYFqITVIb1LgQVd5Ok5EO8pyaSNL5scVspMs8dEWrGWU6Y4zgkB7uPlGtZyrIEaiFE6sNnmckcAayMDuG08GjPeynDBBdcH-UI-8gEDHC9hISDCIpVPikjfEKA__&Key-Pair-Id=APKAIVAVKHB6SNHJAJQQ (3m14.974743446s)
2021/08/16 13:31:35 <-- context canceled PATCH https://gcr.io/v2/imjasonh/elastic/blobs/uploads/AJAD5v13to2CaoVvJcqVMaSZAgX0_m-7JxbvTqOwVnpqXoYjxW1AuRFJrd0BF951w9cs6Jcq9S94F5zexVrgeAs (3m9.176955557s)
2021/08/16 13:31:35 <-- context canceled PATCH https://gcr.io/v2/imjasonh/elastic/blobs/uploads/AJAD5v0slw5ZfvoyBwoEEeokSgVxbCxUU7qMp3VHb1YaJXP4ZQPD1uuoTh-oZq7a6OdkQwpjXiRv4h_AcT4q7x4 (3m14.925109378s)
Error: failed to copy image: Patch "https://gcr.io/v2/imjasonh/elastic/blobs/uploads/AJAD5v2lhllJ7p8DZ3y0TwZ6IV66iLJEPoDSkXPUzimmdaOBK_GprROAINc-RIGDgT-D14iEf4sSgw2eU1EKz5A": context canceled

```

**Comment by Roshick** [comment]
> I'm guessing you have some proxy/firewall preventing you from talking to elastic, and docker [has been configured](https://docs.docker.com/network/proxy/) to use this proxy (or the daemon is running on a separate machine that can access elastic?).

Good hint, I looked in our firewall and could see the requests coming from our proxy going through the firewall, but nothing that would block them. I'm using the crane container and it neither works on the machine where I tried docker (had the container run there with docker) nor our k8s.

I'll try it on my personal computer tomorrow so I can make sure it's not something related to our corporate infrastructure.
